devices/src/virtio/vsock/unix/muxer.rs

```diff
@@ -0,0 +1,736 @@
+                    self.host_sock.accept().map(|_| 0).unwrap_or(0);
+                    return;
+                }
+                self.host_sock
+                    .accept()
+                    .map_err(Error::UnixAccept)
+                    .and_then(|(stream, _)| {
+                        stream
+                            .set_nonblocking(true)
+                            .map(|_| stream)
+                            .map_err(Error::UnixAccept)
+                    })
+                    .and_then(|stream| {
+                        // Before forwarding this connection to a listening AF_VSOCK socket on the
+                        // guest side, we need to know the destination port. We'll read that port
+                        // from a "connect" command received on this socket, so the next step is to
+                        // ask to be notified the moment we can read from it.
+                        self.add_listener(stream.as_raw_fd(), EpollListener::LocalStream(stream))
+                    })
+                    .unwrap_or_else(|err| {
+                        warn!("vsock: unable to accept local connection: {:?}", err);
+                    });
+            }
+
+            // Data is ready to be read from a host-initiated connection. That would be the "connect"
+            // command that we're expecting.
+            Some(EpollListener::LocalStream(_)) => {
+                if let Some(EpollListener::LocalStream(mut stream)) = self.remove_listener(fd) {
+                    Self::read_local_stream_port(&mut stream)
+                        .and_then(|peer_port| Ok((self.allocate_local_port(), peer_port)))
+                        .and_then(|(local_port, peer_port)| {
+                            self.add_connection(
+                                ConnMapKey {
+                                    local_port,
+                                    peer_port,
+                                },
+                                MuxerConnection::new_local_init(
+                                    stream,
+                                    uapi::VSOCK_HOST_CID,
+                                    self.cid,
+                                    local_port,
+                                    peer_port,
+                                ),
+                            )
+                        })
+                        .unwrap_or_else(|err| {
+                            info!("vsock: error adding local-init connection: {:?}", err);
+                        })
+                }
+            }
+
+            _ => {
+                info!("vsock: unexpected event: fd={:?}, evset={:?}", fd, evset);
+            }
+        }
+    }
+
+    /// Parse a host "connect" command, and extract the destination vsock port.
+    ///
+    fn read_local_stream_port(stream: &mut UnixStream) -> Result<u32> {
+        let mut buf = [0u8; 32];
+        stream.read(&mut buf).map_err(Error::UnixRead)?;
```

Comment:
Is there any possibility of reading more bytes than just the `connect <port>` pseudo-command? If so, what happens to those extra bytes? 

Conversely, is it possible to read less than the intended number of bytes? For example, the initiator wants to do `connect 1234`, but we only read `connect 123`. Should we add a terminator to be on the safe side?

Reply:
Yeah, this was a can of worms. Replaced it with a byte-by-byte read to EOL.